File: src/main/java/org/tobi29/scapes/block/ItemStack.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.block;

import org.tobi29.scapes.engine.utils.io.tag.TagStructure;
import org.tobi29.scapes.engine.utils.math.FastMath;

import java.util.Optional;

public class ItemStack {
    private final GameRegistry registry;
    private int amount;
    private int data;
    private Material material;
    private TagStructure metaData;

    public ItemStack(ItemStack item) {
        this(item.material, item.data, item.amount, item.metaData.copy());
    }

    public ItemStack(GameRegistry registry) {
        this(registry.air(), 0, 0);
    }

    public ItemStack(Material material, int data, int amount,
            TagStructure metaData) {
        registry = material.registry();
        this.material = material;
        this.data = data;
        this.amount = amount;
        this.metaData = metaData;
    }

    public ItemStack(Material material, int data) {
        this(material, data, 1);
    }

    public ItemStack(Material material, int data, int amount) {
        this(material, data, amount, new TagStructure());
    }

    public ItemStack(Material material, int data, TagStructure metaData) {
        this(material, data, 1, metaData);
    }

    public int amount() {
        return amount;
    }

    public ItemStack setAmount(int amount) {
        this.amount = amount;
        if (this.amount <= 0 || material == registry.air()) {
            metaData = new TagStructure();
            this.amount = 0;
            material = registry.air();
        }
        return this;
    }

    public int data() {
        return data;
    }

    public ItemStack setData(int data) {
        this.data = data;
        return this;
    }

    public Material material() {
        return material;
    }

    public ItemStack setMaterial(Material material) {
        if (material == null) {
            material = registry.air();
        }
        this.material = material;
        return this;
    }

    public int canStack(ItemStack add) {
        return canStack(add, FastMath.min(
                FastMath.min(material.maxStackSize(this),
                        add.material.maxStackSize(add)) - amount, add.amount));
    }

    public int canStack(ItemStack add, int amount) {
        if ((add.material != material || add.data != data ||
                amount + this.amount > FastMath.min(material.maxStackSize(this),
                        add.material.maxStackSize(add))) &&
                material != registry.air() && this.amount > 0) {
            return 0;
        }
        return amount;
    }

    public int canTake(ItemStack take) {
        return canTake(take, take.amount);
    }

    public int canTake(ItemStack take, int amount) {
        if (take.material != material ||
                take.data != data || material == registry.air() ||
                this.amount <= 0) {
            return 0;
        }
        return FastMath.min(this.amount, amount);
    }

    public void load(TagStructure tag) {
        material = registry.material(tag.getInteger("Type"))
                .orElse(registry.air());
        data = tag.getShort("Data");
        amount = tag.getInteger("Amount");
        metaData = tag.getStructure("MetaData");
    }

    public TagStructure save() {
        TagStructure tag = new TagStructure();
        tag.setInteger("Type", material.itemID());
        tag.setInteger("Data", data);
        tag.setInteger("Amount", amount);
        tag.setStructure("MetaData", metaData);
        return tag;
    }

    public TagStructure metaData(String category) {
        return metaData.getStructure(category);
    }

    public void setMetaData(String category, TagStructure metaData) {
        this.metaData.setStructure(category, metaData);
    }

    public int stack(ItemStack add) {
        return stack(add, FastMath.min(FastMath.min(material.maxStackSize(this),
                add.material.maxStackSize(add)) - amount, add.amount));
    }

    public int stack(ItemStack add, int amount) {
        if ((add.material != material || add.data != data ||
                amount + this.amount > FastMath.min(material.maxStackSize(this),
                        add.material.maxStackSize(add))) &&
                material != registry.air() && this.amount > 0) {
            return 0;
        }
        material = add.material;
        data = add.data;
        if (this.amount == 0) {
            metaData = add.metaData.copy();
        }
        this.amount += amount;
        return amount;
    }

    public Optional<ItemStack> take(ItemStack take) {
        return take(take, take.amount);
    }

    public Optional<ItemStack> take(ItemStack take, int amount) {
        if (take.material != material ||
                take.data != data || material == registry.air() ||
                this.amount <= 0) {
            return Optional.empty();
        }
        ItemStack give = new ItemStack(this);
        give.setAmount(FastMath.min(this.amount, amount));
        this.amount -= give.amount;
        if (this.amount <= 0) {
            metaData = new TagStructure();
            this.amount = 0;
            material = registry.air();
        }
        return Optional.of(give);
    }

    public Optional<ItemStack> take(int amount) {
        if (material == registry.air() || this.amount <= 0) {
            return Optional.empty();
        }
        ItemStack give = new ItemStack(this);
        give.setAmount(amount);
        this.amount -= amount;
        if (this.amount <= 0 || material == registry.air()) {
            metaData = new TagStructure();
            this.amount = 0;
            material = registry.air();
        }
        return Optional.of(give);
    }

    public boolean isEmpty() {
        return amount <= 0 || material == registry.air();
    }

    public String name() {
        return material.name(this);
    }
}

File: ScapesInventory/src/main/kotlin/org/tobi29/scapes/inventory/ItemStack.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.inventory

import org.tobi29.io.tag.TagMap
import org.tobi29.io.tag.toInt
import org.tobi29.io.tag.toTag
import org.tobi29.stdex.math.ceilToInt
import kotlin.math.min

typealias ItemStackable = TypedItem<ItemTypeStackable>
typealias ItemTypeStackable = ItemTypeStackableI<*>

interface ItemTypeStackableI<in I : ItemType> : ItemTypeI<I> {
    /**
     * Stacks one item onto another
     *
     * The implementation shall take all items of bottom and as many as possible
     * from top and return them as the left stack
     *
     * Leftovers of [top] (if any) shall be returned on the right stack
     *
     * As per convention, the `"Amount"` tag has to be updated accordingly
     */
    fun stack(bottom: TypedItem<I>,
              top: ItemStackable,
              max: Int = Int.MAX_VALUE): Pair<Item?, Item?>

    /**
     * Materializes a stack by taking it from another
     *
     * The implementation shall take as many items of top to recreate bottom
     * into the middle stack
     *
     * Leftovers of [top] (if any) shall be returned on the left stack
     *
     * Leftovers of [bottom] (if any) shall be returned on the right stack
     *
     * As per convention, the `"Amount"` tag has to be updated accordingly
     */
    fun take(top: TypedItem<I>,
             bottom: ItemStackable): Triple<Item?, Item?, Item?>
}

interface ItemTypeStackableDefaultI<in I : ItemTypeStackableDefaultI<I>> : ItemTypeStackableI<I> {
    fun maxStackSize(item: TypedItem<I>): Int

    override fun stack(bottom: TypedItem<I>,
                       top: ItemStackable,
                       max: Int): Pair<Item?, Item?> {
        if (top.type != this
                || (bottom.metaData - "Amount") != (top.metaData - "Amount"))
            return bottom to top

        @Suppress("UNCHECKED_CAST")
        val topM = top as TypedItem<I>

        val amount = bottom.amount + topM.amount
        val maxBottom = bottom.type.maxStackSize(bottom)
        val maxTop = bottom.type.maxStackSize(topM)
        val leftAmount = amount.coerceAtMost(max)
                .coerceAtMost(min(maxBottom, maxTop))

        return bottom.copy(amount = leftAmount).orNull() to
                topM.copy(amount = amount - leftAmount).orNull()
    }

    override fun take(top: TypedItem<I>,
                      bottom: ItemStackable): Triple<Item?, Item?, Item?> {
        if (bottom.type != this) return Triple(top, null, bottom)

        val amount = min(top.amount, bottom.amount)
        val (left, right) = top.split(amount)
        return Triple(right, left,
                bottom.copy(amount = bottom.amount - amount).orNull())
    }
}

inline val ItemStackable?.amount: Int
    get() = if (this == null) 0 else metaData["Amount"]?.toInt() ?: 1

fun <T : ItemTypeStackable> TypedItem<T>.copy(
        type: T = this.type,
        amount: Int = this.amount,
        metaData: TagMap = this.metaData
): TypedItem<T> = TypedItem(type,
        (metaData + ("Amount" to amount.toTag())).toTag())

fun <T : ItemTypeStackable> ItemStack(
        type: T,
        amount: Int,
        metaData: TagMap = TagMap()
): TypedItem<T> = TypedItem(type,
        (metaData + ("Amount" to amount.toTag())).toTag())

fun <T, K : Any> ItemStack(
        type: T,
        kind: K,
        amount: Int,
        metaData: TagMap = TagMap()
): TypedItem<T> where T : ItemTypeStackable, T : ItemTypeKinds<K> =
        ItemStack(type, amount, metaData).copy(kind)

fun Item?.split(ratio: Double): Pair<Item?, Item?> =
        kind<ItemTypeStackable>()?.run {
            split((amount * ratio).ceilToInt())
        } ?: this to null

fun Item?.split(left: Int): Pair<Item?, Item?> =
        kind<ItemTypeStackable>()?.run {
            left.coerceAtMost(amount).let {
                copy(amount = it).orNull() to copy(
                        amount = amount - it).orNull()
            }
        } ?: this to null

fun Item?.stack(
        add: Item?,
        max: Int = Int.MAX_VALUE
): Pair<Item?, Item?> = add.kind<ItemTypeStackable>()?.let { other ->
    kind<ItemTypeStackable>()?.run {
        @Suppress("UNCHECKED_CAST")
        (type as ItemTypeStackableI<ItemType>).stack(this, other, max)
    } ?: add to null
} ?: this to add

fun Item?.stackAll(
        add: Item?
): Pair<Item?, Item?> {
    val result = stack(add)
    return if (result.second.isEmpty()) result
    else this to add
}

fun Item?.take(
        take: Item?
): Triple<Item?, Item?, Item?> = take.kind<ItemTypeStackable>()?.let { other ->
    kind<ItemTypeStackable>()?.run {
        @Suppress("UNCHECKED_CAST")
        (type as ItemTypeStackableI<ItemType>).take(this, other)
    }
} ?: Triple(null, this, take)

fun Item?.takeAll(
        take: Item?
): Pair<Item?, Item?> {
    val result = take(take)
    return if (result.third.isEmpty()) result.first to result.second
    else result.first to null
}
